### ISSUE TITLE
Add hints in deps.json to indicate an assembly has been filtered at publish time

### ISSUE THREAD

**Issue by ramarag** [dotnet/sdk#1042]
cc @livarcocc @nguerrera @eerhardt @gkhanna79 

We should decide if the hint should be added per assembly/package or just a global hint to indicate the app has undergone the filtering process at publish time.

This is to enable host to give a better error when such assemblies are not found
https://github.com/dotnet/core-setup/issues/1840

**Comment by gkhanna79** [comment]
Is there a reason why filtered assemblies be treated specially when missing from the cache? IMHO, we should keep the behaviour consistent with the scenario of an app that was incorrectly published/built and attempted to be executed with missing artifacts.

CC @bleroy

**Comment by livarcocc** [comment]
@gkhanna79 I feel like this is special. For instance, I publish filtering based on the aspnet runtime store but then try to execute it in a machine without it.

I believe it is easier for the user to understand that he is missing the shared framework than it would be for the user to understand he is missing the runtime store. If we could help guide the user to solve the problem, that would be great.

cc @bleroy @DamianEdwards

**Comment by bleroy** [comment]
Seems at first like the min-bar would be to add filtering as a possible cause for the missing assembly in the error message. This would not be unreasonable if the user running the app is the same who published it since they are necessarily aware of the filtering. If it's not the same person however, as would happen for example with deployments of pre-built versions of an app, then that would be insufficient. In that case, integrating the name of the target manifest may become the min-bar. I'm not sure what scenario would require the assembly/package level information though: the information about which assemblies failed to load is already there, and if we have the name of the manifest in addition to that, it should be easy to make the connection and properly diagnose what went wrong. It would only be confusing if assembly loading happened for a multitude of reasons, but how often will that happen, and would it really be that much more difficult to handle one reason at a time?
In summary, I'd say a global hint would be enough.

**Comment by gkhanna79** [comment]
> I publish filtering based on the aspnet runtime store but then try to execute it in a machine without it

I get this - however, this scenario is no different from that of an application that is misconfigured. That is, app.deps.json has an entry for the dependency but it is not available in the probing paths. 

> Seems at first like the min-bar would be to add filtering as a possible cause for the missing assembly in the error message.

The host does give an error message when something is missing in probe locations but present in app.deps.json. I think it is reasonable to augment that message to indicate that filtering could be a cause.

**Comment by bleroy** [comment]
It's different in that we anticipate that mistake to be somewhat more common, which could justify setting the bar a little higher, but I'm mostly fine with a generic error message.

**Comment by gkhanna79** [comment]
> if the hint should be added per assembly/package

FWIW, if this were to be done, I would suggest it to be for each entry for the package that was filtered out to ensure that error message can be obvious.

**Comment by bleroy** [comment]
The value of the hint should be the file name without extension of the target manifest that was used to publish the application and that filtered the assembly out.

**Comment by eerhardt** [comment]
I sent this to @ramarag through email, but then realized I should have posted it here.

My proposal for what the addition to the .deps.json file looks like:

Since we filter out whole libraries/NuGet packages, I was thinking something along the lines of adding a new, optional property in the “libraries” section:
```
  "libraries": {
    "dotnet.deps.json/2.0.0-preview1-005685": {
      "type": "project",
      "serviceable": false,
      "sha512": "",
      "runtimeStoreManifestName": "Microsoft.Asp.Net"        <---- new optional property
    },
```

The host can read this new property, and if set, use the name to provide a better error message than "FileNotFound".

**Comment by dsplaisted** [comment]
Can there be multiple runtime stores used at once?

**Comment by ramarag** [comment]
sure the infrastructure we have built allows for the multiple runtime stores
